src/Stratis.Bitcoin.Features.PoA/FederationManager.cs

```diff
@@ -20,45 +22,91 @@ public class FederationManager
         {
             this.settings = Guard.NotNull(nodeSettings, nameof(nodeSettings));
             this.network = Guard.NotNull(network as PoANetwork, nameof(network));
+            this.keyValueRepo = Guard.NotNull(keyValueRepo, nameof(keyValueRepo));
 
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
         }
 
+        // TODO
+        /*
+         * Subscribe to VotingManager and track when new member is added or deleted. Update keys and then persist.
+         */
+
         public void Initialize()
         {
-            this.LoadKey();
+            // Load federation from the db.
+            this.federationMembers = this.LoadFederationKeys();
 
-            if (this.FederationMemberKey != null)
+            if (this.federationMembers == null)
             {
-                // Loaded key has to be a key for current federation.
-                if (!this.network.ConsensusOptions.FederationPublicKeys.Contains(this.FederationMemberKey.PubKey))
-                {
-                    string message = "Key provided is not registered on the network!";
+                this.logger.LogDebug("Federation members are not stored in the db. Loading genesis federation members.");
 
-                    this.logger.LogCritical(message);
-                    throw new Exception(message);
-                }
+                this.federationMembers = this.network.ConsensusOptions.GenesisFederationPublicKeys;
 
-                this.logger.LogInformation("Federation key pair was successfully loaded. Your public key is: {0}.", this.FederationMemberKey.PubKey);
+                this.SaveFederationKeys(this.federationMembers);
             }
 
+            // Display federation.
             this.logger.LogInformation("Federation contains {0} members. Their public keys are: {1}",
-                this.network.ConsensusOptions.FederationPublicKeys.Count, Environment.NewLine + string.Join(Environment.NewLine, this.network.ConsensusOptions.FederationPublicKeys));
+                this.federationMembers.Count, Environment.NewLine + string.Join(Environment.NewLine, this.federationMembers));
+
+            // Load key.
+            Key key = new KeyTool(this.settings.DataFolder).LoadPrivateKey();
+
+            this.IsFederationMember = key != null;
+            this.FederationMemberKey = key;
+
+            if (this.FederationMemberKey == null)
+            {
+                this.logger.LogTrace("(-)[NOT_FED_MEMBER]");
+                return;
+            }
+
+            // Loaded key has to be a key for current federation.
+            if (!this.federationMembers.Contains(this.FederationMemberKey.PubKey))
+            {
+                string message = "Key provided is not registered on the network!";
+
+                this.logger.LogCritical(message);
+                throw new Exception(message);
```

Comment:
Maybe throwing is not the best idea?
that means a federation member cannot start a node until they are approved.

Reply:
Maybe it too complicated to add the member while node is running when approved so no big deal.

Reply:
Yeah, this is smth to prevent running a node thinking you are a fed member when that's not the case. 

I don't think we need to support cases when we start a node with invalid key and during the run we get approved and then enable miner and other components that only fed members require

Reply:
No its more about a fed member that can run a node while they wait to be approved.